src/server/protocol.ts

```diff
@@ -2956,6 +2956,12 @@ namespace ts.server.protocol {
         Smart = "Smart",
     }
 
+    export enum SemicolonPreference {

```

Comment:
Some of the other formatting settings have an auto value that attempts to infer an appropriate behavior from other text in the file. Would that make sense for semicolons?

Reply:
Some of the _user preferences_ have “auto” settings, but none of the formatter settings do, as far as I’ve seen. The distinction is pretty confusing and my guess is the user preferences that deal with formatting arose from the limitations of the formatter. (Before this PR, the formatter dealt _exclusively_ with whitespace.) The distinction isn’t _really_ surfaced to users, so it may not matter much, but it would be nice if we had a more unified story around where this kind of functionality lives. As such, I thought about it, but I’m not sure a _formatter_ is the correct place for inference. That said, when writing new nodes (like auto imports), if you’ve chosen “ignore” (the default), we do make an attempt to match your current file contents. The idea is that when we generate new nodes, we _have_ to make a decision somehow: either the text is going to have semicolons or it’s not.

There’s theoretically nothing stopping us from introducing an “auto” setting for the formatter too, but I’d rather wait and see if people ask for it, purely because I wanted to avoid creating an unnecessarily confusing amount of choices.

Reply:
My only concern with the wait-and-see approach is that VS has a per-user setting for this (vs per-project) so there's no way to disable it if the current TS version doesn't support "auto".  That means that we need to have a good default on the server and it's not clear to me what it would be - ignore?

Reply:
VS Code’s setting is also per-user. The default is “ignore,” because that’s the pre-PR behavior. “Auto” would definitely be more disruptive than “ignore.” Am I understanding your concern correctly?

Reply:
That sounds right to me - the default setting should be to ignore any changes to semicolon formatting.

Reply:
Sorry, "default" was confusing in that context.  What would the fallback setting be if the user specified "auto" and the server didn't support it?

Reply:
Oh, got it. The fallback would be “ignore.” There are a number of preferences/rules that have a disclaimer about what version of TypeScript they require. It’s not the end of the world, but it’s not ideal.

It just gives me an uncomfortable squirmy feeling thinking about an option that means, “I want you to fix up all my semicolons, but I _don’t_ want to tell you how I like them!”

Of course, that has a highly practical application: “Please adapt my semicolon habits to match whatever project I happen to be working on right now.” But, that request may as well apply to _every_ formatter option, not just semicolon preference. As such, I think this is the wrong venue to try to solve the problem of “I work on a bunch of different projects with different formatting conventions.”

Reply:
I agree that it would make more sense to add "auto" everywhere at once and I don't think we're ready to do that, so I'm fine with omitting "auto" for now.